chef/cookbooks/cinder/recipes/ceph.rb

```diff
@@ -17,51 +17,83 @@
+    else
+      Chef::Log.info("Using external ceph cluster for cinder #{volume[:backend_name]} backend, with no automatic setup.")
+      next
+    end
+
+    cmd = ["ceph", "-k", admin_keyring, "-c", ceph_conf, "-s"]
+    check_ceph = Mixlib::ShellOut.new(cmd)
+
+    unless check_ceph.run_command.stdout.match("(HEALTH_OK|HEALTH_WARN)")
+      Chef::Log.info("Ceph cluster is not healthy; skipping the ceph setup for cinder #{volume[:backend_name]} backend")
+      next
+    end
+
+    backend_id = "backend-#{volume[:backend_driver]}-#{volid}"
+
+    cinder_user = volume[:rbd][:user]
+    cinder_pool = volume[:rbd][:pool]
+
+    ceph_clients[ceph_conf] = {} unless ceph_clients[ceph_conf]
+    ceph_keyrings[ceph_conf] = admin_keyring unless ceph_keyrings[ceph_conf]
 
-    check_ceph = Mixlib::ShellOut.new("ceph -k #{volume['rbd']['admin_keyring']} -c #{volume['rbd']['config_file']} -s | grep -q -e 'HEALTH_[OK|WARN]'")
-    check_ceph.run_command
+    cinder_pools = (ceph_clients[ceph_conf][cinder_user] || []) << cinder_pool
+    ceph_clients[ceph_conf][cinder_user] = cinder_pools
 
-    if check_ceph.exitstatus == 0
+    ceph_pool cinder_pool do
+      ceph_conf ceph_conf
+      admin_keyring admin_keyring
+    end
+  end
+end
+
+unless ceph_clients.empty?
+  glance_servers = search(:node, "roles:glance-server")
+  if glance_servers.length > 0
+    glance_pool = glance_servers[0][:glance][:rbd][:store_pool]
+  else
+    glance_pool = nil
+  end
 
-      backend_id = "backend-#{volume['backend_driver']}-#{volid}"
-      cinder_user = volume[:rbd][:user]
-      cinder_pool = volume[:rbd][:pool]
+  ceph_clients.each do |ceph_conf, ceph_pools|
+    ceph_hash = Digest::MD5.hexdigest(ceph_conf)
 
-      cinder_pools = []
-      cinder_pools << cinder_pool
+    ceph_pools.each_pair do |cinder_user, cinder_pools|
       cinder_pools << glance_pool unless glance_pool.nil?
+
       allow_pools = cinder_pools.map{|p| "allow rwx pool=#{p}"}.join(", ")
       ceph_caps = { 'mon' => 'allow r', 'osd' => "allow class-read object_prefix rbd_children, #{allow_pools}" }
 
+      # We have to compute MD5 hash from ceph_conf and cinder_user
+      # to be sure that keyring file will not be overwritten 
+      # by config from another cluster, with same user but different 
+      # config file
+      ceph_hash = Digest::MD5.hexdigest(ceph_conf)
+
       ceph_client cinder_user do
-        unless volume['rbd']['use_crowbar']
-          ceph_conf  volume['rbd']['config_file']
-          admin_keyring  volume['rbd']['admin_keyring']
-        end
+        ceph_conf  ceph_conf
+        admin_keyring  ceph_keyrings[ceph_conf]
         caps ceph_caps
-        keyname "client.#{cinder_user}"
-        filename "/etc/ceph/ceph.client.#{cinder_user}.keyring"
+        keyname "client.#{ceph_hash}.#{cinder_user}"
+        filename "/etc/ceph/ceph.client.#{ceph_hash}.#{cinder_user}.keyring"
```

Comment:
Replying to myself:
> Would it be possible to keep `client.#{cinder_user}` and instead write the file to `/etc/ceph/#{ceph_hash}/ceph.client.#{cinder_user}.keyring`?
Looking at the cinder backend, I don't think that would work :/

I can three options:
a) leave it as it is
b) revert to the old way, and add a check in crowbar to make sure that there will be no conflict for user names
c) add a magic option when there are ceph backends to "ensure there's no conflict on user names"